src/Diagnostics.DataProviders/KustoSDKClient.cs

```diff
@@ -117,7 +129,86 @@ public async Task<DataTable> ExecuteQueryAsync(string query, string cluster, str
+                        cluster = GetBackupClusterName(cluster);
+                    }
+
+                    DiagnosticsETWProvider.Instance.LogRetryAttemptMessage(
+                        requestId ?? string.Empty,
+                        source,
+                        $"Starting Attempt : {attempt}, Cluster: {cluster}, Query : {query}"
+                        );
+
+                    invocationStartTime = DateTime.UtcNow;
+                    attemptException = null;
+                    executeQueryTask = ExecuteQueryAsync(query, cluster, database, DataProviderConstants.DefaultTimeoutInSeconds, requestId, operationName);
+                    dtResult = await executeQueryTask;
+                }
+                catch (Exception ex)
+                {
+                    attemptException = ex;
+                    exceptions.Add(ex);
+                }
+                finally
+                {
+                    invocationEndTime = DateTime.UtcNow;
+                    var totalResponseTime = invocationEndTime - invocationStartTime;
+                    string exceptionType = attemptException != null ? attemptException.GetType().ToString() : string.Empty;
+                    string exceptionDetails = attemptException != null ? attemptException.ToString() : string.Empty;
+
+                    DiagnosticsETWProvider.Instance.LogRetryAttemptSummary(
+                            requestId ?? string.Empty,
+                            source,
+                            $"Attempt : {attempt},  IsSuccessful : {attemptException == null}, Cluster: {cluster}, Query : {query}",
+                            Convert.ToInt64(totalResponseTime.TotalMilliseconds),
+                            invocationStartTime.ToString("HH:mm:ss.fff"),
+                            invocationEndTime.ToString("HH:mm:ss.fff"),
+                            exceptionType,
+                            exceptionDetails
+                            );
+
+                    // Logic to check if retry needs to continue
+                    if(totalResponseTime.TotalSeconds > _config.MaxFailureResponseTimeInSecondsForRetry || !IsExceptionRetryable(attemptException))
+                    {
+                        DiagnosticsETWProvider.Instance.LogRetryAttemptMessage(
+                        requestId ?? string.Empty,
+                        source,
+                        $"Not continuing Retries after Attempt : {attempt}, Cluster: {cluster}, Query : {query}"
+                        );
+
+                        attempt = _config.MaxRetryCount;
+                    }
+
+                    attempt++;
+                }
+
+                if (attempt <= _config.MaxRetryCount) await Task.Delay(_config.RetryDelayInSeconds * 1000);
+
+            } while (attempt <= _config.MaxRetryCount);
+
+            if(executeQueryTask.IsCompleted && !executeQueryTask.IsFaulted && !executeQueryTask.IsCanceled)
```

Comment:
>executeQueryTask.IsCompleted && !executeQueryTask.IsFaulted && !executeQueryTask.IsCanceled [](start = 15, length = 91)

You can use Task.IsCompletedSuccessfully
[https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.iscompletedsuccessfully?view=netcore-3.1](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.iscompletedsuccessfully?view=netcore-3.1)